packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt

```diff
@@ -3603,18 +3603,41 @@ actual object RealmInterop {
         }
 
         override fun notify(work_queue: CPointer<realm_work_queue_t>?) {
-            scope.launch {
-                try {
-                    printlntid("on dispatcher")
-                    lock.withLock {
-                        if (!cancelled) {
-                            realm_wrapper.realm_scheduler_perform_work(work_queue)
+            try {
+                scope.launch {
+                    try {
+                        printlntid("on dispatcher")
+                        lock.withLock {
+                            if (!cancelled) {
+                                realm_wrapper.realm_scheduler_perform_work(work_queue)
+                            }
                         }
+                    } catch (e: Exception) {
+                        // Should never happen, but is included for development to get some indicators
+                        // on errors instead of silent crashes.
+                        e.printStackTrace()
                     }
-                } catch (e: Exception) {
-                    // Should never happen, but is included for development to get some indicators
-                    // on errors instead of silent crashes.
-                    e.printStackTrace()
+                }
+            } catch (ex: IllegalStateException) {
+                // A work-around for https://github.com/realm/realm-kotlin/issues/1608
```

Comment:
I guess we have had some debate on which way to organize the dispatcher/scheduler relationship (https://github.com/realm/realm-kotlin/pull/1559 and https://github.com/realm/realm-kotlin/pull/1463#pullrequestreview-1551509397), but this seems like a very unsafe way of guarding this. We already have infrastructure in place for being able to cancel the `SingleThreadDispatcherScheduler`, so seems way safer to expose the `cancel` to the upper layers, so that we follow the exact logic on when we expect this to happen (when we close the realm?). I guess that wouldn't work if people inject their own dispatcher and just close it, but we could document that they shouldn't close the dispatcher before closing the realm.

Reply:
I modified the implementation, so we now have two locks instead, one to control the dispatcher and one for the scheduler.